nomad/secure_variables_endpoint.go

```diff
@@ -1,133 +1,470 @@
+		// Attempt to lookup AuthToken as a Node.SecretID and return any error
+		// wrapped along with the original.
+		node, stateErr := sv.srv.fsm.State().NodeBySecretID(nil, args.AuthToken)
+		if stateErr != nil {
+			var mErr multierror.Error
+			mErr.Errors = append(mErr.Errors, err, stateErr)
+			return mErr.ErrorOrNil()
+		}
+
+		// At this point, we do not have a valid ACL token, nor are we being
+		// called, or able to confirm via the state store, by a node.
+		if node == nil {
+			return structs.ErrTokenNotFound
+		}
+	}
+
+	// Update via Raft.
+	out, index, err := sv.srv.raftApply(structs.SecureVariableDeleteRequestType, args)
+	if err != nil {
 		return err
-	} else if aclObj != nil && !aclObj.IsManagement() {
-		return structs.ErrPermissionDenied
 	}
 
-	// TODO: implementation
-	SV_List(args, reply)
+	// Check if the FSM response, which is an interface, contains an error.
+	if err, ok := out.(error); ok && err != nil {
+		return err
+	}
 
+	// Update the index. There is no need to floor this as we are writing to
+	// state and therefore will get a non-zero index response.
+	reply.Index = index
 	return nil
 }
 
+// GetNamespace is used to get a specific namespace
 func (sv *SecureVariables) Read(args *structs.SecureVariablesReadRequest, reply *structs.SecureVariablesReadResponse) error {
-	if done, err := sv.srv.forward("SecureVariables.Read", args, args, reply); done {
+	if done, err := sv.srv.forward(structs.SecureVariablesReadRPCMethod, args, args, reply); done {
 		return err
 	}
-
 	defer metrics.MeasureSince([]string{"nomad", "secure_variables", "read"}, time.Now())
 
-	// TODO: implement real ACL checks
-	if aclObj, err := sv.srv.ResolveToken(args.AuthToken); err != nil {
+	// Perform our mixed auth handling.
+	if err := sv.handleMixedAuthEndpoint(args.QueryOptions, acl.NamespaceCapabilityReadJob); err != nil {
 		return err
-	} else if aclObj != nil && !aclObj.IsManagement() {
-		return structs.ErrPermissionDenied
 	}
 
-	// TODO: implementation
-	SV_Read(args, reply)
+	// Setup the blocking query
+	opts := blockingOptions{
+		queryOpts: &args.QueryOptions,
+		queryMeta: &reply.QueryMeta,
+		run: func(ws memdb.WatchSet, s *state.StateStore) error {
+			// Look for the namespace
+			out, err := s.GetSecureVariable(ws, args.RequestNamespace(), args.Path)
+			if err != nil {
+				return err
+			}
+
+			// Setup the output
+			reply.Data = nil
+			if out != nil {
+				reply.Data = out.Copy()
+				if err := sv.decrypt(reply.Data); err != nil {
+					return err
+				}
+				reply.Index = out.ModifyIndex
+			} else {
+				// Use the last index that affected the secure table
+				index, err := s.Index(state.TableSecureVariables)
+				if err != nil {
+					return err
+				}
+
+				// Ensure we never set the index to zero, otherwise a blocking query cannot be used.
+				// We floor the index at one, since realistically the first write must have a higher index.
+				if index == 0 {
+					index = 1
+				}
+				reply.Index = index
```

Comment:
We've got a helper method for this block:

```suggestion
				 sv.srv.replySetIndex(state.TableSecureVariables, &reply.QueryMeta)
```